refactor(form-plugin): remove duplicate code + add tests

arturovt opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
```
[ ] Bugfix
[ ] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

arturovt commented:
@markwhitfeld @eranshmil @splincode FYI

I've refactored duplicate code, like 4 calls to `markForCheck`. Also excluded double subscription that will cost less memory. Removed subscription inside subscription (that is bad in concepts of rxjs). Also added additional tests

eranshmil commented on tsconfig.json:
```diff
@@ -10,6 +10,7 @@
     },
     "strict": true,
     "strictPropertyInitialization": false,
+    "suppressImplicitAnyIndexErrors": true,
```
What's the reason behind this configuration?

arturovt replied:
@eranshmil forgot to remove :wink:

arturovt commented:
@markwhitfeld @eranshmil @splincode FYI

I've refactored duplicate code, like 4 calls to markForCheck and unnecessary nested if's that can be replaced with filter operator. Also excluded double subscription that will cost less memory. Removed subscription inside subscription (that is bad in concepts of rxjs). Also added additional tests

markwhitfeld commented on packages/form-plugin/src/directive.ts:
```diff
@@ -1,23 +1,25 @@
-import { Subject } from 'rxjs';
-import { takeUntil, debounceTime, first } from 'rxjs/operators';
+import { Subject, merge, pipe, iif, of } from 'rxjs';
+import { takeUntil, debounceTime, filter, tap, mergeMap, finalize, map } from 'rxjs/operators';
```
This is a much bigger footprint of rxjs operators. See comments below.

markwhitfeld commented:
Sorry 😢

splincode commented:
@markwhitfeld Well, before that, the code looked really awful, and I’m sure it’s more difficult to maintain it, because it’s not decomposed at all. In the books of Steve McConnell there are completely different practices that are used in this project and this makes me sad.
